pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -845,32 +858,25 @@ func (h *upgradeHandler) storeVMState(upgrade *harvesterv1.Upgrade) error {
 		"namespace": upgrade.Namespace,
 		"name":      upgrade.Name,
 	}
-	isSingleNodeRestoreVM, err := h.isSingleNodeRestoreVM(upgrade)
-	if err != nil {
-		return err
-	}
-	if !isSingleNodeRestoreVM {
-		logrus.WithFields(logFields).Info("Skip store VM state")
-		return nil
-	}
 
 	if upgrade.Annotations != nil && upgrade.Annotations[preUpgradeVMsAnnotation] != "" {
 		logrus.WithFields(logFields).Infof("Skip store VM state since %s is already set", preUpgradeVMsAnnotation)
 		return nil
 	}
 
-	vms, err := h.vmCache.List(corev1.NamespaceAll, labels.Everything())
+	vmis, err := h.vmiCache.List(corev1.NamespaceAll, labels.Everything())
 	if err != nil {
 		logrus.WithFields(logFields).WithError(err).Error("Failed to list all VMs")
 		return err
 	}
 
 	runningVMs := make([]*PreUpgradeRunningVM, 0)
-	for _, vm := range vms {
+	for _, vm := range vmis {
```

Comment:
just save all VMs ?

Reply:
Not sure if I missed anything-would appreciate any hints you can share! 😃  
BTW @ihcsim left some suggestion in https://github.com/harvester/harvester/pull/8138#discussion_r2067418652, perhaps this is something you want to point out?

Reply:
It was explained at above; and was similar to @ihcsim 's suggestion; Harvester needs to book exactly the VM operated by it and restore upon them.   The failure case https://github.com/harvester/harvester/issues/8212 is also thoughtful.

>A main complex case is:

>The running VMs info are all saved as pre-upgrade-vm, and when a node is upgraded, no matter successful/failed, start all none-running VMs originally hosted on this node.

>I did not find any code to save the VMs info explicitly when the VMs are turned off by upgrade code. If a running VM is turned off by user in-between the upgrade due do various of reasons, the VM may still be powered on accidentally by the new code. (please correct me if I am wrong). Could this be improved?
```